net/nss/files/patch-ppc64.diff

```diff
@@ -0,0 +1,11 @@
+--- nss/lib/freebl/mpi/mpcpucache.c.orig	1970-01-01 08:00:00.000000000 +0800
++++ nss/lib/freebl/mpi/mpcpucache.c	2022-05-12 21:28:37.000000000 +0800
+@@ -705,7 +705,7 @@
+ #define MPI_GET_PROCESSOR_LINE_SIZE_DEFINED 1
+ #endif
+ 
+-#if defined(__ppc64__)
++#if !defined(__APPLE__) && defined(__ppc64__)
```

Comment:
just disabling this code might he OK….as you say, ppc gets by without it. But it looks like this code was supposed to work, reading through the code upstream…

Reply:
> just disabling this code might he OK….as you say, ppc gets by without it. But it looks like this code was supposed to work, reading through the code upstream…

Apparently the upstream did not have in mind that Darwin supported `ppc64`. Otherwise they would have had a case for `ppc` there too, which they do not. IMO it is quite obviously intended for ELF, and that’s it fails on Leopard for `ppc64`.

Reply:
I don’t see anything ELF in there… or relating to any ABI at all, at least in this one I found browsing:

https://chromium.googlesource.com/chromium/deps/nss/+/ccb083050bac653bd6b98c38237fdf93c5d64a7a/nss/lib/freebl/mpi/mpcpucache.c#695

and Apple is at least mentioned at one point…

Reply:
so it looks to me like it was supposed to work. Exactly what happens when you try to compile this? It was not clear from the ticket what went wrong.

Reply:
> so it looks to me like it was supposed to work. Exactly what happens when you try to compile this? It was not clear from the ticket what went wrong.

This:
```
:info:build mpi/mpcpucache.c: In function ‘dcbzl’:
:info:build mpi/mpcpucache.c:730: error: nested functions are disabled, use -fnested-functions to re-enable
:info:build mpi/mpcpucache.c:730: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘asm’
:info:build mpi/mpcpucache.c:730: warning: implicit declaration of function ‘asm’
:info:build mpi/mpcpucache.c:730: error: lvalue required as left operand of assignment
:info:build mpi/mpcpucache.c:732: error: ‘a’ undeclared (first use in this function)
:info:build mpi/mpcpucache.c:732: error: (Each undeclared identifier is reported only once
:info:build mpi/mpcpucache.c:732: error: for each function it appears in.)
:info:build mpi/mpcpucache.c:731: error: invalid lvalue in asm output 0
:info:build make[4]: *** [Output.OBJD/Darwin_SINGLE_SHLIB/mpcpucache.o] Error 1
:info:build make[4]: Leaving directory `/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_net_nss/nss/work/nss-3.78-ppc64/nss/lib/freebl'
:info:build make[3]: *** [freebl_FREEBL_BUILD_SINGLE_SHLIB] Error 2
```

Reply:
anyway, if no processor-specific method works the code falls back to a generic default value, so no big deal if Apple ppc64 doesn’t use an optimal method I suppose, even if it did compile and work.

Reply:
> I don’t see anything ELF in there… or relating to any ABI at all, at least in this one I found browsing:
> 
> https://chromium.googlesource.com/chromium/deps/nss/+/ccb083050bac653bd6b98c38237fdf93c5d64a7a/nss/lib/freebl/mpi/mpcpucache.c#695
> 
> and mac is at least mentioned at one point…

Yes, I have read the relevant source code, of course. In fact, there is an instruction in Darwin ABI `dcbzl128` which may be what they would want, but their code fails, as a matter of fact, on Leopard. My assumption re ELF is based on this: I mean, Leopard did not change, so if the code was working back then, it will work now too. But it does not.

Reply:
> anyway, if no processor-specific method works the code falls back to a generic default value, so no big deal if Apple ppc64 doesn’t use an optimal method I suppose, even if it did compile and work.

Indeed. So then I just remove a platform restriction for the patch now, and we are good to go with this?

Reply:
so this little bit is what fails to compile? I should compile on Darwin:



/* clear the cache line that contains 'array' */
--
  | static inline void dcbzl(char *array)
  | {
  | register char *a asm("r2") = array;
  | __asm__ __volatile__( "dcbzl %0,r0" : "=r" (a): "0"(a) );
  | }

Reply:
> so this little bit is what fails to compile? It should compile on Darwin:
> 
> /* clear the cache line that contains 'array' */   static inline void dcbzl(char *array)   {   register char *a asm("r2") = array;   **asm** **volatile**( "dcbzl %0,r0" : "=r" (a): "0"(a) );   }

Thank you! I can check this once I am back to the Quad. No G5 machine at home.

Reply:
@kencu By the way, maybe you could suggest how to change this line in `libuv` code for Darwin `ppc64`? https://github.com/libuv/libuv/issues/3623#issuecomment-1124859031

Reply:
upstream has rewritten this function like this now:

https://github.com/nss-dev/nss/blob/671d89b6c4a6f41707bb044534751098e2e3f211/lib/freebl/mpi/mpcpucache.c#L725


```
/* clear the cache line that contains 'array' */
static inline void
dcbzl(char *array)
{
    register char *a asm("r2") = array;
    __asm__ __volatile__("dcbzl %0,0"
                         : "=r"(a)
                         : "0"(a));
}
```

which is pretty similar.

anyway if it won't compile, it won't I guess. No big loss to fall back to the (slower) default fallback.

Reply:
I had trouble getting it to compile on godbolt until I removed the "static"

https://godbolt.org/z/esG4Gj7TP